### PR TITLE
[DevOps] Create the Bot Framework CLI Task 

### PR DESCRIPTION
**Pull request by Bill7zz**
### Description
This pull request adds a DevOps task that uses the features of the [BotFramework CLI](https://docs.microsoft.com/en-us/azure/bot-service/bf-cli-overview?view=azure-bot-service-4.0).

The task is a wrapper of the principal CLI commands:

- ✔️ [LUIS](https://www.luis.ai/) 
- ✔️ [QnA Maker](https://www.qnamaker.ai/)
- Dispatch 

The task consists of getting the user input and run the BF-Cli tools using the inputs as arguments. 
image

The next images show the BF-Cli task running the commands as we expected.
image

### CODE REVIEW

File: build/botframework-cli-task-publish.yml [hunk 1]
```diff
@@ -0,0 +1,76 @@
...
+
+trigger: none
+pr: none
+
+jobs:
+  - job: BotFramework_CLI_Task_Publish
+    variables:
+      buildVersion: '4.8.0-preview.$(Build.BuildId)'
+      _version: ${{coalesce(variables.version, variables.buildVersion)}}
+      publisherId: 'Southworks-Pipelines-Test'
```
Comment (ceciliaavila) [inline build/botframework-cli-task-publish.yml:19]: Please change the publisher Id to 'Microsoft'

File: build/botframework-cli-task-publish.yml [hunk 2]
```diff
@@ -0,0 +1,76 @@
...
+        publisherId: '$(publisherId)'
+        extensionName: '$(extensionName)'
+        extensionVersion: '$(version)'
+        extensionVisibility: private
+        extensionPricing: free
+
+    - task: ms-devlabs.vsts-developer-tools-build-tasks.publish-extension-build-task.PublishAzureDevOpsExtension@2
+      displayName: 'Publish Extension'
+      inputs:
+        connectedServiceName: MarketPlaceConnection
```
Comment (ceciliaavila) [inline build/botframework-cli-task-publish.yml:69]: This should be a variable.

File: packages/bfCliTask/src/qnaMakerCommand.ts [hunk 3]
```diff
@@ -0,0 +1,332 @@
...
+                    this.publishKnowledgeBase();
+                    break;
+                case 'ReplaceKB':
+                    this.replaceKnowledgeBase();
+                    break;
+                case 'UpdateKB':
+                    this.updateKnowledgeBase();
+                    break;
+                case 'TrainKB':
+                    this.traingKnowledgeBase();
```
Comment (ceciliaavila) [inline packages/bfCliTask/src/qnaMakerCommand.ts]: Here there's a typo. 
```suggestion
                    this.trainKnowledgeBase();
```

File: packages/bfCliTask/src/qnaMakerCommand.ts [hunk 4]
```diff
@@ -0,0 +1,332 @@
...
+        } catch (error) {
+            taskLibrary.setResult(taskLibrary.TaskResult.Failed, error.message, true);
+        }
+
+    }
+
+    private publishKnowledgeBase = async (newKbId?: string) => {
+        console.log('Publishing QnA knowledgebase');
+    
+        let command = `bf qnamaker:kb:publish --kbId "${ newKbId? newKbId: this.kbId }" --subscriptionKey "${ this.qnaKey }"`;
```
Comment (ceciliaavila) [inline packages/bfCliTask/src/qnaMakerCommand.ts:117]: this could be a const.

File: packages/bfCliTask/src/qnaMakerCommand.ts [hunk 5]
```diff
@@ -0,0 +1,332 @@
...
+        if (this.publishNewKB) {
+            let ObjKbId = JSON.parse(stdout.toString());
+            this.publishKnowledgeBase(ObjKbId.kbId);
+        }
+    }
+    
+    private deleteKnowledgeBase = async () => {
+        console.log('Deleting QnA knowledgebase');
+
+        let command = `bf qnamaker:kb:delete --kbId "${ this.kbId }" --subscriptionKey "${ this.qnaKey }" --force`;
```
Comment (ceciliaavila) [inline packages/bfCliTask/src/qnaMakerCommand.ts:154]: This should be a const.

File: packages/bfCliTask/vss-extension.json [hunk 6]
```diff
@@ -0,0 +1,44 @@
+{
+    "manifestVersion": 1,
+     "id": "bf-cli-task",
+     "name": "BotFramework Task",
+     "version": "4.8.0.82",
+     "publisher": "Southworks-Pipelines-Test",
```
Comment (ceciliaavila) [inline packages/bfCliTask/vss-extension.json:6]: Replace this with 'Microsoft'

**Review by ceciliaavila — CHANGES_REQUESTED** [review]

File: packages/bfCliTask/src/task.json [hunk 7]
```diff
@@ -0,0 +1,557 @@
...
+            }
+        },
+        {
+            "name": "qnaKey",
+            "type": "string",
+            "label": "QnA Subscription Key",
+            "required": true,
+            "groupname": "command",
+            "visibleRule": "qnaMakerCommand = TRUE && qnaMakerSubCommand!=QnAConvert && qnaMakerSubCommand!=QnATranslate",
+            "helpMarkDown": "Enter you QnAMaker found on the Cognitive Services Azure portal page under 'access keys'"
```
Comment (ceciliaavila) [inline packages/bfCliTask/src/task.json:350]:
```suggestion
            "helpMarkDown": "Enter your QnAMaker key found on the Cognitive Services Azure portal page under 'access keys'"
```

File: packages/bfCliTask/src/task.json [hunk 8]
```diff
@@ -0,0 +1,557 @@
...
+        },
+        {
+            "name": "kbName",
+            "type": "string",
+            "label": "Knowledgebase name",
+            "required": false,
+            "defaultValue": "",
+            "groupname": "command",
+            "visibleRule": "qnaMakerSubCommand=CreateKB || qnaMakerSubCommand=QnAConvert",
+            "helpMarkDown": "Enter a name for your knowledgebase"
```
Comment (ceciliaavila) [inline packages/bfCliTask/src/task.json:360]:
```suggestion
            "helpMarkDown": "Enter a name for your knowledge base"   
```

File: packages/bfCliTask/src/task.json [hunk 9]
```diff
@@ -0,0 +1,557 @@
...
+            "helpMarkDown": "Enter a name for your knowledgebase"   
+        },
+        {
+            "name": "kbId",
+            "type": "string",
+            "label": "Knowledgebase ID",
+            "required": true,
+            "groupname": "command",
+            "visibleRule": "qnaMakerSubCommand=QueryKB || qnaMakerSubCommand=DeleteKB || qnaMakerSubCommand=PublishKB || qnaMakerSubCommand=ReplaceKB || qnaMakerSubCommand=UpdateKB || qnaMakerSubCommand=TrainKB",
+            "helpMarkDown": "Enter the ID of the knowledgebase."
```
Comment (ceciliaavila) [inline packages/bfCliTask/src/task.json:369]:
```suggestion
            "helpMarkDown": "Enter the ID of the knowledge base."
```

File: packages/bfCliTask/src/task.json [hunk 10]
```diff
@@ -0,0 +1,557 @@
...
+            "required": false,
+            "defaultValue": "",
+            "groupname": "command",
+            "visibleRule": "qnaMakerSubCommand=CreateKB || qnaMakerSubCommand=QnAConvert",
+            "helpMarkDown": "Enter a name for your knowledgebase"   
+        },
+        {
+            "name": "kbId",
+            "type": "string",
+            "label": "Knowledgebase ID",
```
Comment (ceciliaavila) [inline packages/bfCliTask/src/task.json:365]:
```suggestion
            "label": "Knowledge base ID",
```

File: packages/bfCliTask/src/task.json [hunk 11]
```diff
@@ -0,0 +1,557 @@
...
+            "label": "QnA Subscription Key",
+            "required": true,
+            "groupname": "command",
+            "visibleRule": "qnaMakerCommand = TRUE && qnaMakerSubCommand!=QnAConvert && qnaMakerSubCommand!=QnATranslate",
+            "helpMarkDown": "Enter you QnAMaker found on the Cognitive Services Azure portal page under 'access keys'"
+        },
+        {
+            "name": "kbName",
+            "type": "string",
+            "label": "Knowledgebase name",
```
Comment (ceciliaavila) [inline packages/bfCliTask/src/task.json:355]:
```suggestion
            "label": "Knowledge base name",
```

File: packages/bfCliTask/src/task.json [hunk 12]
```diff
@@ -0,0 +1,557 @@
...
+            "label": "Knowledgebase ID",
+            "required": true,
+            "groupname": "command",
+            "visibleRule": "qnaMakerSubCommand=QueryKB || qnaMakerSubCommand=DeleteKB || qnaMakerSubCommand=PublishKB || qnaMakerSubCommand=ReplaceKB || qnaMakerSubCommand=UpdateKB || qnaMakerSubCommand=TrainKB",
+            "helpMarkDown": "Enter the ID of the knowledgebase."
+        },
+        {
+            "name": "kbDTOFileLocation",
+            "type": "filepath",
+            "label": "Knowledgebase DTO file location",
```
Comment (ceciliaavila) [inline packages/bfCliTask/src/task.json:374]:
```suggestion
            "label": "Knowledge base DTO file location",
```

File: packages/bfCliTask/src/task.json [hunk 13]
```diff
@@ -0,0 +1,557 @@
...
+            "required": false,
+            "groupname": "command",
+            "visibleRule": "qnaMakerSubCommand=CreateKB || qnaMakerSubCommand=ReplaceKB || qnaMakerSubCommand=UpdateKB",
+            "helpMarkDown": "Enter the location of your knowledgebase DTO file"
```
Comment (ceciliaavila) [inline packages/bfCliTask/src/task.json:378]:
```suggestion
            "helpMarkDown": "Enter the location of your knowledge base DTO file"
```

File: packages/bfCliTask/src/task.json [hunk 14]
```diff
@@ -0,0 +1,557 @@
...
+        },
+        {
+            "name": "publishNewKB",
+            "type": "boolean",
+            "label": "Publish knowledgebase",
```
Comment (ceciliaavila) [inline packages/bfCliTask/src/task.json:383]:
```suggestion
            "label": "Publish knowledge base",
```

File: packages/bfCliTask/src/task.json [hunk 15]
```diff
@@ -0,0 +1,557 @@
...
+            "required": false,
+            "groupname": "command",
+            "defaultValue": "false",
+            "visibleRule": "qnaMakerSubCommand=CreateKB",
+            "helpMarkDown": "Check to publish this knowledgebase after the creation."
```
Comment (ceciliaavila) [inline packages/bfCliTask/src/task.json:388]:
```suggestion
            "helpMarkDown": "Check to publish the knowledge base after its creation."
```

File: packages/bfCliTask/src/task.json [hunk 16]
```diff
@@ -0,0 +1,557 @@
...
+        },
+        {
+            "name": "feedbackRecordDTOFileLocation",
+            "type": "filepath",
+            "label": "FeedbackRecord DTO file location",
```
Comment (ceciliaavila) [inline packages/bfCliTask/src/task.json:393]:
```suggestion
            "label": "Feedback Record DTO file location",
```

File: packages/bfCliTask/src/task.json [hunk 17]
```diff
@@ -0,0 +1,557 @@
...
+            "required": false,
+            "groupname": "command",
+            "visibleRule": "qnaMakerSubCommand=TrainKB",
+            "helpMarkDown": "Enter the location of your FeedbackRecord DTO file"
```
Comment (ceciliaavila) [inline packages/bfCliTask/src/task.json:397]:
```suggestion
            "helpMarkDown": "Enter the location of your Feedback Record DTO file"
```

File: packages/bfCliTask/src/task.json [hunk 18]
```diff
@@ -0,0 +1,557 @@
...
+        },
+        {
+            "name": "overrideServiceEndpointKB",
+            "type": "boolean",
+            "label": "Override Public Endpoint",
+            "required": false,
+            "groupname": "command",
+            "defaultValue": "false",
+            "visibleRule": "qnaMakerSubCommand=AlterationsReplaceKB",
+            "helpMarkDown": "Supported Cognitive Services endpoints (protocol and hostname, for example: https://westus.api.cognitive.microsoft.com).."
```
Comment (ceciliaavila) [inline packages/bfCliTask/src/task.json:416]:
```suggestion
            "helpMarkDown": "Supported Cognitive Services endpoints (protocol and hostname, for example: https://westus.api.cognitive.microsoft.com)."
```

File: packages/bfCliTask/src/task.json [hunk 19]
```diff
@@ -0,0 +1,557 @@
...
+            "helpMarkDown": "Query to get a prediction for"
+        },
+        {
+            "name": "qnaConvertFilePathInput",
+            "type": "filepath",
+            "label": "Input Files",
+            "defaultValue": "",
+            "groupname": "command",
+            "visibleRule": "qnaMakerSubCommand=QnAConvert",
+            "helpMarkDown": "Source .qna file(s) or QnA KB JSON file to convert. Empty is the root of the repo (build) which is $(System.DefaultWorkingDirectory)"
```
Comment (ceciliaavila) [inline packages/bfCliTask/src/task.json:461]:
```suggestion
            "helpMarkDown": "Source .qna file(s) or QnA KB JSON file to convert. Leave empty for the root of the repo (build) which is $(System.DefaultWorkingDirectory)"
```

File: packages/bfCliTask/src/task.json [hunk 20]
```diff
@@ -0,0 +1,557 @@
...
+        },
+        {
+            "name": "isAlterationFile",
+            "type": "boolean",
+            "label": "Is QnA Alterations file",
+            "required": false,
+            "groupname": "command",
+            "defaultValue": "false",
+            "visibleRule": "qnaMakerSubCommand=QnAConvert",
+            "helpMarkDown": "Indicates if files is QnA Alterations"
```
Comment (ceciliaavila) [inline packages/bfCliTask/src/task.json:471]:
```suggestion
            "helpMarkDown": "Indicates if the file is QnA Alterations"
```

File: packages/bfCliTask/src/task.json [hunk 21]
```diff
@@ -0,0 +1,557 @@
...
+        },
+        {
+            "name": "qnaConvertFilePathOutput",
+            "type": "filepath",
+            "label": "Output Files",
+            "required": false,
+            "groupname": "command",
+            "visibleRule": "qnaMakerSubCommand=QnAConvert",
+            "helpMarkDown": "Output file or folder name. If not specified stdout will be used as output. Empty is the root of the repo (build) which is $(System.DefaultWorkingDirectory)"
```
Comment (ceciliaavila) [inline packages/bfCliTask/src/task.json:480]: This is kind of confusing. not specified = empty?

File: packages/bfCliTask/src/task.json [hunk 22]
```diff
@@ -0,0 +1,557 @@
...
+            "helpMarkDown": "Output folder name. If not specified stdout will be used as output"
+        },
+        {
+            "name": "qnaTranslateKey",
+            "type": "string",
+            "label": "Translation Key",
+            "required": false,
+            "groupname": "command",
+            "visibleRule": "qnaMakerSubCommand=QnATranslate",
+            "helpMarkDown": "Machine translation endpoint key"
```
Comment (ceciliaavila) [inline packages/bfCliTask/src/task.json:507]:
```suggestion
            "helpMarkDown": "The key of the Translation Text service endpoint."
```

**Review by ceciliaavila — CHANGES_REQUESTED** [review]

File: build/botframework-cli-task-publish.yml (hunk 2 above)
Comment (Bill7zz) [reply build/botframework-cli-task-publish.yml:69]: Updated to `'$(MarketPlaceConnection)'`

**Pull request closed without merging**